#186: add file-transfer UX brief; close three open product questions (#196)
khmelevartem/tether: docs/product/features/file-transfer/ux-brief.md

Author: подумай, может будет полезно на этих двух стейтах выделять карточку подсветкой или заливкой? или достаточно, что у нас рисуется прогресс?

Author: Решил **не** добавлять отдельную подсветку. Аргументы:<br>• Swell (рост карточки) уже даёт явный визуальный cue "это та карточка где сейчас активность".<br>• Brand-mark line filling в accent color даёт второй визуальный сигнал.<br>• Background tint конкурировал бы с accent-цветом brand-mark.<br>• При concurrent transfers (2+) все активные карточки swell-нутые — общая подсветка стала бы шумом, индивидуальная — не отличала бы их между собой.<br><br>Если на проде увидим, что пользователи теряют активную карточку среди других в длинном списке — добавим subtle surface elevation (тень/тонкий border) как progressive enhancement; пока стартуем со swell + brand-mark fill, не утяжеляем визуал.

                                                                  │
                                                  (PeerCard → Active outbound)
      │
      └── PeerCard Cancelled / Error / Received (partial) / Sent (partial)
                                                                  │
                                                       [Show received files →]
                                                                  │
                                                    TransferDetailsScreen

SettingsSection — File Transfer  (inside the app's existing settings surface)
```

Screens introduced: LargeSelectionConfirmDialog, TransferDetailsScreen, MobilePickerChooserSheet, SettingsSection — File Transfer.

Screens touched: DeviceListScreen (peer rows replaced by PeerCards with full state machine).

---

## Screens

### DeviceListScreen

**Purpose.** The home screen where the user sees online peers (as PeerCards) and manages any in-flight or pending transfer state.

**Entry points.** App launch; OS share-sheet / "Open with" routing Tether to foreground with files pre-selected; drag-and-drop files onto the Tether window (macOS/Desktop).

**Layout.**

> - Top bar: app title only.

Author: ```suggestion
```

1. **Receiver-side retry for batch failures.** Sender-side retry is in scope for this feature; receiver-initiated retry is deferred. A receiver-side retry — where the receiver itself requests the sender to re-send only the failed files after freeing space — requires a pull-protocol shape, a decision on how long the failed-batch reference persists, and whether the sender must still be online. Deferred until the sender-side retry ships and usage signals the demand.

2. **Mobile picker unification — the "two taps" gap.** The MobilePickerChooserSheet adds one tap to the vision's "two taps to send" on Android and iOS, caused by OS constraints (the OS file/folder pickers do not support mixing folder and multi-file selection in a single picker session). Revisit once OS support evolves or an in-app picker becomes viable.

3. **iOS deep-link to Files app reliability.** The inline hint in PeerCard Received state ("Open Files → On My iPhone → Tether") is a static instruction when the deep-link fails. If the deep-link proves unreliable in practice, a more guided in-app flow may be needed. Monitor failure rate post-launch.

> 4. **Receiver-cancel copy symmetry.** Flow 7 uses "Received \<X\> files from \<peer\> — you cancelled." for the receiver's own Received partial-completion state. An alternative is "…— transfer ended early." which is less accusatory. Confirm preferred framing before implementation.

Author: ```suggestion
4. **Receiver-cancel copy symmetry.** Flow 7 uses "Received \<X\> files from \<peer\> — cancelled by reciever." for the receiver's own Received partial-completion state. An alternative is "…— transfer ended early." which is less accusatory. Confirm preferred framing before implementation.
```

Author: Подумал и оставил "you cancelled". Аргумент: с self-perspective ресивера "cancelled by receiver" звучит как 3rd-person self-reference (странно). "Transfer ended early" — эвфемизм, скрывает агента. "You cancelled" — прямая фактическая формулировка собственного действия; не accusatory, а neutral acknowledgment. Tether's tone — calm/clear, не извинительно-эвфемистичный. Если "accusatory" для тебя критично — переключусь на "transfer ended early", но я бы аргументировал за прямоту.

# UX brief — File transfer

**Spec:** [spec.md](spec.md)
**Status:** `ready`

---

## Information architecture

This feature introduces six new screens / dialogs, touches one existing screen (DeviceListScreen), and adds one settings section.

```
DeviceListScreen  (existing — touched)
├── [banner: pending outbound]   ← AutoPickInlinePrompt (first-time only, at top)
├── [banner: active outbound]
├── IncomingTransferCard(s)      ← stacked, inline, above device rows
└── [tap peer] ──────────────────────────────────────┐
                                                      ▼
                              MobilePickerChooserSheet (Android + iOS only)
                                    │
                              [large folder?] ──── FolderSendConfirmDialog
                                    │
                              SendProgressScreen
                                    │
                              TransferSummaryScreen
                                         │ (iOS notification tap)

>                               ReceivedScreen (iOS only, notification-tap entry)

Author: не очень понял, почему это айос только?

Author: ReceivedScreen удалён вообще — на всех платформах теперь Received-стейт PeerCard.

AutoPickConfirmDialog  (modal, fired inline when toggle flipped ON)

SettingsSection — File Transfer  (inside the app's existing settings surface)
```

Screens introduced: SendProgressScreen, IncomingTransferCard, FolderSendConfirmDialog, TransferSummaryScreen, AutoPickInlinePrompt, AutoPickConfirmDialog, ReceivedScreen (iOS only), MobilePickerChooserSheet, SettingsSection — File Transfer.

Screens touched: DeviceListScreen.

---

## Screens

### DeviceListScreen

**Purpose.** The home screen where the user sees online peers and manages any in-flight or pending transfer state.

**Entry points.** App launch; OS share-sheet / "Open with" routing Tether to foreground with files pre-selected; return from SendProgressScreen or TransferSummaryScreen.

**Layout.**

> - Top bar: app title, settings affordance.

Author: кажется, это вне скоупа. убери, чтобы не сбивать исполнителя. будет отдельно

Author: Settings affordance убрана из top bar — out of scope.

- AutoPickInlinePrompt strip (first-time only, see that screen) — rendered between the banner (if any) and the device rows.
- Scrollable list of device rows (searching / empty / populated — per existing spec).
- IncomingTransferCard(s) — stacked, rendered above device rows when inbound transfers exist.

**States.**

- **Searching (no peers yet, no pending transfer):** the `•—•` brand mark in its searching state (hollow right dot, opacity oscillation); copy: "Searching for devices…". No banner.
- **Searching (with pending outbound):** same searching animation; banner above reads "Ready to send \<N\> files (\<size\>). Pick a device below." with [Cancel] on the right.
- **Populated (no pending transfer):** device rows visible. No banner.
- **Populated (pending outbound):** device rows visible; banner reads "Ready to send \<N\> files (\<size\>). Pick a device below." with [Cancel].
- **Active outbound (transfer in flight from this device):** banner reads "Sending to \<peer\> — \<N\> files, \<progress\>%". [Cancel] on right. Device rows still visible beneath.
- **Active inbound (one transfer arriving):** IncomingTransferCard rendered above device rows.

> - **Active inbound (two concurrent transfers):** two stacked IncomingTransferCards.

Author: продумай, оцени с точки зрения UX, как будет лучше. мб инлайнить этот стейт в карточку пира? то есть не создавать отдельную карточку, а расширять уже имеющуюся?

Author: Решено: PeerCard как универсальный контейнер. Active inbound — состояние карточки пира, отдельной IncomingTransferCard больше нет.

- **Banner success toast (4 s, after outbound completes):** banner transitions to success-tone: "Sent to \<peer\>", then self-dismisses.

> - **Banner cancelled toast (4 s, after outbound cancelled):** banner transitions to neutral-tone: "Cancelled", then self-dismisses.

Author: я бы не делал self-dismisses. сценарий - пользователь поставил отправку, пока ждет - отошел. когда вернется, захочет явно увидеть, что передача состоялась, а уже все стерлось. нужен другой триггер исчезновения баннера
по первому - точно, по второму -наверное

Author: Снят везде — Sent/Received/Cancelled/Error персистентны до явного [Dismiss ×]. Pending banner — до выбора пира или [Cancel].

- **iOS foreground constraint banner (during any transfer):** persistent "Keep Tether open" banner rendered above the pending/active banner (or at the very top if no transfer banner exists). Copy: "Keep Tether open to complete the transfer." Not dismissible. Disappears when no transfer is active.

**Interactions.**

> - Tap pending-banner [Cancel]: clears pending selection, dismisses banner (no confirm dialog). If a transfer was active, both sides stop immediately.

Author: [Cancel] - это имеется в виду кнопка? это не просто понять. нужно более явное указание

речь же про оба баннера - pending/active? уточни

Author: Везде в layout-описаниях явно "[Cancel button]" / "[Retry button]" / "[Done button]" — однозначно кнопка. Применено к pending и active banner.

---

### MobilePickerChooserSheet

**Purpose.** A bottom sheet on Android and iOS that lets the user choose which system picker to invoke — Photos, Files, or Folder — before a send flow begins.

**Entry points.** Tapping a device row on Android or iOS when no pending outbound exists (in-app send). Also reachable if a pending outbound was initiated in-app and the user needs to add more files (not in MVP scope — single batch only).

**Layout.**

- Bottom-sheet handle at top.
- Sheet title: "Send from…"
- Three options as tappable rows with icons: "Photos", "Files", "Folder".
- Swipe down or tap outside: dismisses without action.

**States.**

- **Default:** three options listed, no loading state.
- **Dismissed:** sheet gone, user returns to DeviceListScreen with no pending state.

**Interactions.**

- Tap "Photos": opens system photo picker (multi-select). On completion, sets pending outbound state and returns to DeviceListScreen with banner. Then navigates to SendProgressScreen after user taps the target device (or auto-picks if toggle ON).
- Tap "Files": opens system file picker (multi-select). Same on completion.

> - Tap "Folder": opens system folder picker (single selection). Same on completion. If selected folder exceeds the soft threshold (>500 files OR >2 GB), FolderSendConfirmDialog appears before proceeding.

Author: предупреждение распространяется на все варианты

Author: LargeSelectionConfirmDialog (переименован) — триггерится для любого выбора >500 файлов / >2 GB: folder picker, multi-file picker, share-sheet, drag-and-drop.

- iOS: "Photos" maps to PHPickerViewController; "Files" and "Folder" map to UIDocumentPickerViewController (in the appropriate mode).
- macOS: this screen does not exist — system file dialog handles both files and folder (separate "Open" and "Open Folder" actions on the same dialog, or two distinct affordances on the SendProgressScreen entry).
- Desktop JVM: this screen does not exist — same as macOS.

**Accessibility.**

- Bottom sheet is presented as a modal; focus moves into the sheet on open.
- Each option row has a semantic label matching its visible label ("Photos", "Files", "Folder").
- Dismiss affordance (handle, scrim): semantic label "Dismiss picker chooser".
- Focus returns to the triggering device row on dismiss.

---

### FolderSendConfirmDialog

**Purpose.** A confirmation dialog that appears when the selected folder exceeds the soft threshold (>500 files OR >2 GB), guarding against an accidental "select all" scenario.

**Entry points.** After folder selection (from MobilePickerChooserSheet on Android/iOS, or system file dialog on macOS/Desktop), when the selection exceeds the threshold.

> **Layout.**

Author: добавь галочку "не показывать больше"

Author: Добавлен checkbox "Don't show again" в LargeSelectionConfirmDialog.

**Copy.**

- "Large folder"
- "About to send \<N\> files (\<size\>) to \<peer\>. Continue?"
- "Cancel"
- "Send"

**Per-platform deltas.**

- Android: system-style dialog. Hardware back = Cancel.
- iOS: UIAlertController-style presentation. Swipe-to-dismiss blocked (destructive intent — default on Cancel keeps Cancel as default).
- macOS: standard sheet attached to the app window.
- Desktop JVM: standard modal dialog.

**Accessibility.**

- Dialog role: `alertdialog`.
- On open, focus is placed on [Cancel].
- [Cancel] semantic label: "Cancel — discard folder selection".
- [Send] semantic label: "Send \<N\> files to \<peer\>".
- Escape key always triggers Cancel on Desktop/macOS.

---

### SendProgressScreen

**Purpose.** The sender's primary screen during an active transfer, showing batch progress via the `•—•` brand mark and live speed.

**Entry points.** Immediately after pairing (or silently if already paired) when the receiver accepts the connection. Replaces the DeviceListScreen on the navigation stack (sender navigates here; back is suppressed during transfer). On iOS, this screen is also the anchor for the foreground constraint banner.

**Layout.**

> - Top bar: peer name as title; [Cancel] affordance in the trailing position.

Author: чем обусловлено расположение кнопки отмены именно здесь?

Author: Cancel теперь живёт на самой PeerCard (на том, что представляет передачу), не в отдельном баре. Justification — карточка и есть transfer surface.

- Center region: the `•—•` brand mark in transfer-progress state — the connecting line fills accent color left-to-right proportional to total bytes transferred in the batch. The mark is the only progress bar; no secondary bar.
- Below the mark: current filename (one line, center-truncated with ellipsis in the middle when too long).

> - Below filename: current speed (e.g. "3.2 MB/s").

Author: а как же "передано Х из У МБ"?

Author: Добавлено: "X.X MB of Y.Y MB" + speed везде на transfer-progress surface'ах (Active outbound / Active inbound / SendProgressScreen).

**States.**

- **Preparing (connection established, enumeration in progress):** `•—•` in transfer-progress state at 0%; below the mark: "Preparing…"; speed field: empty.
- **In-progress:** `•—•` filling left-to-right; filename shows current file being sent; speed shows "Calculating..." for the first ~3 s, then live value (e.g. "3.2 MB/s"). Skip badge appears if any files are skipped so far.
- **Completed (all files sent, no failures):** `•—•` in success state (right dot animation fires); screen transitions automatically to TransferSummaryScreen after the ~700 ms success animation completes.
- **Completed (partial failures):** same success animation, then transitions to TransferSummaryScreen showing the partial-failure summary.
- **All failed:** `•—•` in error state (line truncated at failure point, right dot hollow in error tone); screen transitions to TransferSummaryScreen.
- **Connection lost mid-transfer:** `•—•` in error state; overlay or inline message below mark: "Connection lost. Try again when you're back on Wi-Fi."; [Retry] button visible; [Done] secondary.

> - **Peer unreachable mid-transfer:** `•—•` in error state; message: "\<peer\> is no longer reachable. Try again."; [Retry] and [Done].

Author: в стейты добавить ожидание: когда соединение прекратилось (не завершилось штатным образом), ждем некоторое время, если в течение таймаута восстанавливается - то молча продолжаем передачу. если нет, то переходим к состоянию ошибки.

Author: Добавлен новый стейт "Connection paused / reconnecting" — 15s окно, fresh countdown на каждый drop, silent resume; по timeout → Error.

- **Cancelled (by sender):** brief neutral-tone "Cancelled" message on this screen, then returns to DeviceListScreen (no TransferSummaryScreen).
- **Cancelled (by receiver):** same — sender's screen shows neutral "Cancelled", returns to DeviceListScreen.

**Interactions.**

- Tap [Cancel]: immediately cancels transfer; no confirm dialog. Both sides stop. Screen shows "Cancelled" briefly then returns to DeviceListScreen.

> - Tap [Retry] (error state): re-initiates the same transfer (all files or failed files depending on the error type) to the same peer if peer is reachable.

Author: какие типы ошибок могут быть? я бы ожидал, что ретрай всегда пытается отправить только еще не переданные файлы

Author: Универсальное правило: retry всегда повторяет только не подтверждённые на receiver файлы. Добавлена error matrix (network lost / peer gone / file unreadable / receiver write failed / all failed) — для каждого что повторяется и retry availability.

- Tap [Done] (error state): returns to DeviceListScreen without retry.

> - Hardware back (Android) / swipe-back gesture (iOS): suppressed during active transfer (prevents accidental cancel). If transfer is in an error state, back is permitted (equivalent to [Done]).

Author: нет, я бы сделал по другому. нажатие назад позволительно. можно даже добавить кнопку-стрелку вниз для сворачивания этого экрана до баннера на экран пиров. клик по нему обратно разворачивает экран передачи. нужно явно различить в этом случае отмену и сворачивание экрана

Author: Back = minimize на SendProgressScreen (плюс явная стрелка ▾). Cancel — отдельная destructive-кнопка. Тап PeerCard на DeviceListScreen разворачивает SendProgressScreen обратно. Distinct affordances.

**Layout.**

- Card region (stacked above device rows, one card per concurrent sender):
  - Sender name (e.g. "From Artem's Phone")
  - Subtitle: "Receiving \<N\> files (\<total size\>)"
  - `•—•` brand mark in transfer-progress state (line fills left-to-right as bytes arrive)
  - [Cancel] affordance in trailing position on the card

**States.**

- **Receiving (in-progress):** card visible; `•—•` filling; subtitle shows file count and total size.
- **Success (4-second hold):** `•—•` transitions to success state; card background shifts to success tone; copy changes to "Received \<N\> files from \<peer\> — tap to open". After 4 seconds, card self-dismisses.
- **Cancelled (by receiver or sender):** card disappears immediately; a brief neutral-tone toast appears: "Transfer from \<peer\> cancelled".
- **Error / connection lost:** card switches to error state; `•—•` in error state; copy: "Transfer from \<peer\> interrupted. Ask \<peer\> to retry."; card persists until dismissed by the user (no auto-dismiss on error — user needs to read it).

> - **iOS suspension / screen lock interruption:** when the OS suspends Tether mid-transfer on iOS, on next foreground a one-time informational card appears (distinct from an active card): "Transfer from \<peer\> was interrupted. Ask \<peer\> to send again." with a [Dismiss] affordance.

Author: предусмотрен ли у нас где то механизм, позволяющий устройств не блокироваться (засыпать) во время блокировки? если нет, то надо дописать в #195

Author: Добавлен раздел Sleep & wake lock: Android FGS уже держит wake-lock; macOS — OS sleep-prevention assertion; iOS — невозможно (foreground-only). Sender-side android wake-lock parity вынесен в Open UX questions с указанием на #195.

- iOS: tap success card → opens ReceivedScreen (in-app). OS completion notification fires when possible; tapping notification also opens ReceivedScreen. iOS suspension card (see iOS interruption state above) is iOS-only.
- macOS: tap success card → Tether activates and reveals and selects the received file in Finder. OS notification also fires with the same effect.
- Desktop JVM: tap success card → opens File Explorer (Windows) or best-effort file manager (Linux) at the parent folder. OS notification fires with the same effect. Linux reveal may silently no-op in some DEs — the card tap is the user's primary affordance.

**Accessibility.**

- Each card is a live region (polite); announces once on arrival: "Receiving \<N\> files from \<peer\>".
- [Cancel] semantic label: "Cancel incoming transfer from \<peer\>".
- Success state card: role changes to button; semantic label: "Received \<N\> files from \<peer\>. Tap to open."; announces once when state changes (assertive, once only).
- Error/interruption state: assertive announcement: "Transfer from \<peer\> interrupted."
- Two stacked cards: each card is independently focusable.

---

> ### TransferSummaryScreen

Author: с одной стороны, здесь достаточно текста для вынесения в отдельный абзац.

с другой стороны, это скорее часть (особый стейт) SendProgressScreen, чем отдельный экран. подумай, как это выразить явно

Author: ReceivedScreen полностью устранён, его функционал — Received state карточки пира (на всех платформах). Не отдельный экран.

- "Show details"
- "Hide details"
- "Retry failed files"
- "Retrying…"
- "\<peer\> is no longer reachable."
- "Done"

**Per-platform deltas.**

- Android: hardware back returns to DeviceListScreen (same as [Done]).
- iOS: swipe-back returns to DeviceListScreen. Top bar shows a standard back chevron.
- macOS: top bar shows close affordance (×). Window is not closed — returns to DeviceListScreen within the same window.
- Desktop JVM: same as macOS.

**Accessibility.**

- `•—•` mark `contentDescription`: "Transfer complete" (success) or "Transfer completed with failures" (any failures).
- [Show details] / [Hide details] is a button; semantic label matches the visible label exactly ("Show details" / "Hide details") depending on state.
- [Retry failed files] semantic label when disabled: "Retry not available — \<peer\> is no longer reachable."
- Failed file list: each filename is a read-only text element, not interactive.
- Live-region announcement on screen entry (assertive): the primary summary line is announced once.
- Keyboard focus order (Desktop/macOS): back/close → summary text → expand/collapse button → [Retry failed files] (if present) → [Done].

---

### AutoPickInlinePrompt

> **Purpose.** A one-time inline prompt at the top of the device list that lets the user opt into auto-sending to a single online paired peer, surfaced the first time the condition is met naturally.

Author: я бы сделал не один раз, а каждый раз, когда выполняется условие, и юзер не отказался явно

Author: Toggle переехал в expanded PeerCard и доступен всегда (через шеврон). Auto-discovery исчез — пользователь сам решает когда включить. Срабатывает каждый раз когда условие выполнено и toggle включён.

> **Entry points.** Rendered automatically at the top of the DeviceListScreen (between the pending/active banner and the device rows) the **first time** all three conditions hold simultaneously: a pending outbound exists (share-sheet arrival, drag-and-drop arrival, or the user picked files in-app and is now on the device list with an active pending-banner), exactly one online paired peer is currently visible, and the prompt has never been shown before. Shown once ever (until the user dismisses it or enables the toggle). The prompt does not appear when the user opens Tether with no pending outbound.

Author: по мне, лучше не над, а под единственным пиром

Author: AutoPickInlinePrompt удалён. Toggle теперь внутри expanded PeerCard под строкой пира — не "над/под списком".

**Layout.**

- A dismissible strip / inline card below the pending/active banner (or at the top of the list if no banner is active).

> - Body: "Always send to \<peer\> when it's your only online device?"

Author: изначально, я думал про то, что этот тоггл будет работать на любого пира. но с твоей формулировкой я подумал с другой стороны - может не делать отдельную карточку, а расширить карточку пира и включить в нее?

шеврон - свернуть-развернуть, "auto-send to this device", [i] -> "Always send to \<peer\> when it's your only online device?" toast, toggle

продумай, рассмотри оба варианта - ux-expert.

сохранение тоггла на каждого пира поотдельности мне нравится с той стороны, что пользователь может сам решить, каким пирам "доверять", а каким нет.

Author: Сделано per-peer. Toggle — внутри expanded PeerCard (chevron развёртывает, видна строка "Auto-send to this device when it's your only online device" + [i] tooltip + toggle). Пользователь решает каким пирам доверять.

- **Cancelled ([Cancel]):** toggle reverts to Off; no send initiated; dialog closes; AutoPickInlinePrompt strip remains (toggle shows Off again).

**Interactions.**

- Tap [OK]: save preference, initiate send, close dialog.
- Tap [Cancel]: revert toggle, close dialog.
- Hardware back / Escape: same as [Cancel].

**Copy.**

- "Auto-send enabled"
- "You can turn this off in Settings. Sending to \<peer\> now."
- "Cancel"
- "OK"

**Per-platform deltas.**

- Android: default; hardware back = Cancel.
- iOS: UIAlertController style.
- macOS: standard alert sheet.
- Desktop JVM: standard modal dialog; Escape = Cancel.

**Accessibility.**

- Dialog role: `alertdialog`.
- Default focus: [OK].
- [Cancel] semantic label: "Cancel — keep auto-send off."
- [OK] semantic label: "Enable auto-send and send to \<peer\> now."

---

### ReceivedScreen (iOS only)

**Purpose.** An in-app screen on iOS that confirms a completed inbound transfer and helps the user navigate to the received files, substituting for a direct Files deep-link.

> **Entry points.** Tapping an OS completion notification on iOS (which opens Tether). Also reachable by tapping the success state of IncomingTransferCard before it self-dismisses.

Author: хочется его заинлайнить до особого стейта IncomingTransferCard

1. прием завершен - по тапу на карточку пытаемся пройти по диплинку
2. если диплинк не сработал - показываем внутри этой же карточки подсказку подходящим способом.
3. если нажат пуш, то попадаем либо на девайс скрин с уже показанной подсказкой, либо пытаемся при этом перейти по диплинку - тут подумай сам, как связать

Author: Сделано: Received — состояние PeerCard. По тапу пробует deep-link; если провалился — inline hint внутри той же карточки (persistent до [Dismiss ×]). Push notification ведёт на DeviceListScreen, карточка пира в Received state.

- Tap [Done] or back: returns to DeviceListScreen.

**Copy.**

- "Files received"
- "Received \<N\> files from \<peer\>."
- "Saved to: On My iPhone → Tether/"
- "Show in Files"
- "Open Files → On My iPhone → Tether"
- "Done"

**Per-platform deltas.**

- iOS only — this screen does not exist on Android, macOS, or Desktop.

**Accessibility.**

- `•—•` mark `contentDescription`: "Transfer complete".
- [Show in Files] semantic label: "Open Files app at Tether folder".
- Inline fallback instruction: role is static text; not a button.
- [Done] semantic label: "Done — return to device list".

---

### SettingsSection — File Transfer

**Purpose.** The settings area where the user configures save location and the auto-send toggle.

**Entry points.** App settings (existing settings surface) — a dedicated "File Transfer" section within it.

**Layout.**

- Section header: "File Transfer"
- Save location row: label "Save location"; value shows the current path (e.g. "Downloads/Tether/"); on editable platforms, a disclosure affordance (chevron or "Change" link) opens the system folder picker.

> - Auto-send toggle row: label "Auto-send when only one device is online"; description: "Skip the device list and send immediately when only one of your paired devices is online."; toggle control (On/Off).

Author: если этот тоггла станет per-peer, то здесь дожен быть выпадающий список - https://github.com/khmelevartem/tether/pull/196/changes#r3269537537
или эта секция вообще становится ненужной, если можно редактировать это в карточке пира

Author: Settings секция упрощена: только Save location row. Auto-send toggle удалён оттуда — управляется per-peer в карточке.

**States.**

- **Android (editable):** save location shows current path; tap row → system folder picker to change.
- **iOS (read-only):** save location shows "On My iPhone → Tether/"; no change affordance. A "(fixed)" caption beneath the path: "iOS does not allow changing this location."
- **macOS (editable):** save location shows current path (default: "~/Downloads/Tether/"); disclosure affordance → system folder picker.
- **Desktop JVM (editable):** same as macOS.

> - **Auto-send toggle On:** when enabled, a helper text appears below the row: "Active — will send automatically to \<peer name\> when it's your only online device." (If no peer has been designated yet, helper text: "Will activate when exactly one paired device is online.")

Author: если оставляем вариант - тоггл на всех пиров, то его должно быть можно включать и выключать вне зависимости от выполнения условия 1 пир онлайн. и текст - "will send automatically when single paired device online"

Author: N/A — выбрали per-peer.

8. User taps [Done] → DeviceListScreen.

**Receiver side:**
- IncomingTransferCard success state holds for 4 seconds ("Received \<N\> files from \<peer\> — tap to open"), then self-dismisses.
- OS notification fires. Tapping opens the saved location (platform-appropriate).

### Flow 2 — Share-sheet entry, already paired

1. User is in Photos / Files app. Taps Share → Tether.
2. Tether opens at DeviceListScreen with pending-outbound banner: "Ready to send \<N\> files (\<size\>). Pick a device below."
3. User taps target peer's row.
4. If folder threshold exceeded: FolderSendConfirmDialog. Otherwise: proceeds directly.
5. SendProgressScreen, then TransferSummaryScreen — same as Flow 1 from step 5.

### Flow 3 — Auto-send toggle ON, one peer online (canonical: share-sheet entry)

1. User taps Share → Tether (or drags files onto the Tether window on macOS/Desktop) with auto-send toggle On and exactly one paired peer online.
2. DeviceListScreen is skipped. Pending-outbound banner appears immediately: "Sending to \<peer\>" with [Cancel].
3. Transfer begins without user picking a peer.
4. SendProgressScreen, then TransferSummaryScreen.

> Note: auto-send also fires when the user picks files in-app (via MobilePickerChooserSheet or system file dialog) and returns to DeviceListScreen with a pending outbound — if exactly one peer is online and the toggle is On, the send begins immediately without the user tapping a device row.

Author: не понял, как можно вернуться из MobilePickerChooserSheet в pending состояние. при подтвержденном выборе должна начаться передача. при отмененном - ничего

Author: Уточнено: picker confirmed → передача стартует сразу. Picker cancelled → ничего, остаёмся на DeviceListScreen. Pending state — только при приходе извне (share-sheet / drag-and-drop), не из in-app picker'а.

5. Alternatively, user taps [Cancel] in dialog: toggle reverts to Off, strip remains visible with toggle Off.
6. Alternatively, user taps [×] on strip: strip dismissed permanently, toggle stays Off, no send.

### Flow 5 — Partial batch failure and retry

1. Transfer completes with some file failures (per-file errors during send).
2. SendProgressScreen transitions to TransferSummaryScreen.
3. TransferSummaryScreen shows: "\<X\> of \<Y\> files sent. \<K\> failed." with [Show details] button.
4. User taps [Show details] → list of failed filenames expands inline; button becomes [Hide details].
5. User taps [Retry failed files] (peer still online) → SendProgressScreen re-appears for the failed-file subset.
6. If retry succeeds: TransferSummaryScreen shows "Sent \<K\> files to \<peer\>." (only the retried batch).
7. If peer went offline: [Retry failed files] is disabled; helper text "\<peer\> is no longer reachable" shown below button.

### Flow 6 — Cancel mid-transfer (sender)

1. Transfer is in progress on SendProgressScreen.
2. User taps [Cancel].
3. Both sides stop immediately. No confirm dialog.
4. SendProgressScreen shows brief "Cancelled" then returns to DeviceListScreen.

> 5. Receiver's IncomingTransferCard dismisses with toast: "Transfer from \<peer\> cancelled".

Author: cancell ведь не отменяет тех файлов, что уже передались, так? нужно ли обозначить, какие файлы все таки дошли в этом случае? подумай

Author: Добавлено: [Show received files] / [Hide received files] toggle в Cancelled state — раскрывает inline список фактически дошедших файлов. Copy: "X files were received before cancel. Y were not sent."

### Flow 7 — Cancel mid-transfer (receiver)

1. Inbound transfer is in progress. IncomingTransferCard visible on receiver's DeviceListScreen.
2. Receiver taps [Cancel] on the card.
3. Both sides stop immediately.
4. Card dismisses with toast: "Transfer from \<peer\> cancelled".

> 5. Sender's SendProgressScreen shows neutral "Cancelled", returns to DeviceListScreen.

Author: тот же вопрос про указание уже переданных файлов

Author: То же добавлено для Error state (при partial delivery) и для Received partial-completion на receiver-стороне.

### Flow 8 — Connection lost mid-transfer

1. Transfer is in progress. Wi-Fi drops or peer becomes unreachable.
2. SendProgressScreen switches to error state: `•—•` in error state.
3. If Wi-Fi lost: "Connection lost. Try again when you're back on Wi-Fi." [Retry] [Done].
4. If peer gone: "\<peer\> is no longer reachable. Try again." [Retry] [Done].

> 5. User taps [Retry]: re-initiates transfer to the same peer (if peer is back on mDNS). Returns to in-progress state.

Author: ```suggestion
5. User taps [Retry]: re-initiates transfer to the same peer (if peer is reachable). Returns to in-progress state.
```

Author: Suggestion применён.

12. **Retry-failed-files button** — primary button that is disabled (with peer-gone helper text) when the peer has left mDNS.
13. **Auto-pick inline prompt strip** — one-time dismissible strip with a toggle and [×] affordance; lives at the top of the device list.
14. **Auto-pick confirm dialog** — modal confirmation when auto-send is enabled; default focus OK; immediate send on confirm.
15. **iOS foreground constraint banner** — persistent non-dismissible system-style banner informing the user to keep Tether open during transfers; iOS only.
16. **iOS received screen** — in-app completion screen with "Show in Files" deep-link affordance and inline fallback instruction; iOS only.
17. **Settings save-location row** — editable (with system folder picker disclosure) on Android/macOS/Desktop; read-only with explanatory caption on iOS.
18. **Settings auto-send toggle row** — toggle with contextual helper text describing the current auto-send target peer.
19. **macOS window-close transfer warning** — sheet attached to the window when the user closes Tether mid-transfer; macOS and Desktop only.

---

## Open UX questions

> These are non-blocking future considerations. None gate the current implementation.

Author: добавь. когда мы в состоянии pending - добавить возможность редактирования списка выбранных файлов

Author: Зафиксировано как Open UX question — не в MVP. В pending state редактирования списка нет, выбор замораживается на момент picker'а.

- Escape key always triggers Cancel on Desktop/macOS.

---

### TransferDetailsScreen

**Purpose.** Supplementary per-file view of a transfer — during transit, after full success, after partial failure, after cancel, or after error. Always opt-in via [Show details →] on the PeerCard; the card remains the primary transfer surface.

**Entry points.** Tapping [Show details →] on a PeerCard in any of: Active outbound, Active inbound, Sent (full), Sent (partial), Received (full), Received (partial), Cancelled, or Error.

**Layout.**

- Top bar with back affordance and peer name as title; transfer summary as subtitle (e.g. "Receiving 3 of 8…" while active, "Received 3 of 8 files" once finished).
- Scrollable list of files grouped into sections:
  - "Received" section: files that arrived in full (file name + size).
  - "Sending" section (active states only, sender side): the file currently in flight, plus files queued. Each row shows a small inline progress bar for the active file; queued rows show file name only.
  - "Receiving" section (active states only, receiver side): same as "Sending" from the receiver's perspective.

>   - "Not sent" section (terminal states only): files that did not complete (name only, no size).

Author: взвесь непредвзято, можешь провести небольшое исследование с использованием сабагента и анализом конкурентов: как лучше - группировать "в процессе", "отправленные", "не отправленные" или рисовать напротив каждого файла иконку, отображающую статус?

Author: Запустил research по конкурентам (LocalSend, Telegram, AirDrop, Quick Share, iCloud, Dropbox, Google Drive, Finder/Explorer copy, Send Anywhere) — отчёт под спойлером ниже. Вердикт: **Model B (status icons на стабильных строках)** превалирующий паттерн в категории. Применил.<br><br>**Что меняется:** TransferDetailsScreen теперь — flat list в OS-picker order, position строк стабильна на протяжении всей передачи. Trailing slot каждой строки рендерит один из 5 статусов: Queued (часы) / In progress (inline progress bar) / Done (✓) / Failed (× с inline-helper) / Retrying (визуально как In progress). Sender-side добавлены affordance: [× button] для cancel single file (Queued/In progress) и сама строка становится [Retry button] для Failed.<br><br>**Над списком** — aggregate header strip "X of Y sent · K failed" recomputes live; даёт at-a-glance summary без шумных section headers.<br><br>**Почему Model B победил:**<br>1. **Position predictability** — порядок из OS picker = ментальный индекс пользователя ("третий — contract.pdf"). Sectioned model тасует строки на каждом file completion.<br>2. **Symmetry** — одна layout для всех статусов, нет structural разницы между in-progress и terminal.<br>3. **Scale** — работает одинаково на 3 и на 500 файлах. Sectioned-headers съедают вертикаль на маленьких батчах и создают анимационный шум на больших.<br>4. **Accessibility** — screen reader не теряет focus context (стабильная строка, меняется только value).<br>5. **Industry pattern** — все основные конкуренты (LocalSend, Telegram, iCloud, Dropbox, Google Drive, Quick Share) делают именно так.<br><br><details><summary>Полный research-отчёт</summary><br>**Observed patterns:**<br>• LocalSend, Telegram, iCloud, Dropbox, Google Drive, Quick Share, Send Anywhere — все используют Model B (status icons на стабильных строках в selection order).<br>• AirDrop, WhatsApp, Finder/Explorer copy — aggregate progress без per-file list вообще.<br>• Никто не реорганизует строки по статусу мид-трансфер.<br><br>**Sources:** LocalSend repo + DeepWiki, Telegram Desktop issues (#26266, #26557), AirDrop Apple Community thread, Google Drive Quora help, Dropbox sync-icons docs, iCloud Drive file-status help, WhatsApp media help, Uploadcare UX best practices.</details>

- Empty sections are omitted entirely.

> - Sender-side terminal states with a non-empty "Not sent" section: a [Retry all →] CTA above the list re-sends every file in "Not sent". Per-row [Retry] button on each "Not sent" entry re-sends that single file.

Author: может добавим кнопку отсановить передачу? для состояния на этом экране, когда она еще идет

Author: Сделал. Добавил [Cancel transfer button] в top bar trailing position TransferDetailsScreen во время in-progress mode — destructive-styled, без confirm-диалога (та же семантика что [Cancel button] на PeerCard). Кнопка пропадает когда screen переходит на terminal layout. Адверс-ревью подсветило что нужно явно описать что происходит с Queued/In progress строками при whole-transfer cancel — теперь зафиксировано: они становятся Failed с inline-helper "Transfer cancelled" (отличается от per-file "Cancelled by you"); [Retry] доступен.<br><br>**Бонус:** добавил per-file cancel — [× button] рядом со status-indicator на Queued/In progress строках (только sender-side). Тап → файл становится Failed с "Cancelled by you", остальные файлы в батче продолжают. Receiver-side per-file cancel требует protocol-signal "skip this file" — вынес в Open UX question.

# UX brief — File transfer

**Spec:** [spec.md](spec.md)
**Status:** `ready`

---

## Information architecture

> This feature introduces four new screens / dialogs plus one settings section, touches one existing screen (DeviceListScreen). All per-peer transfer state is contained within PeerCard — there are no separate incoming-transfer cards, auto-pick prompts, or iOS-only received screens.

Author: ```suggestion
This feature introduces four new screens / dialogs plus one settings section, touches one existing screen (DeviceListScreen). All per-peer transfer state is contained within PeerCard.
```

Author: Suggestion применена — IA-вступление теперь именно в этой формулировке.

                                                                  │
                                                     [large selection?] ─── LargeSelectionConfirmDialog
                                                                  │
                                                       SendProgressScreen
                                                                  │
                                                       TransferSummaryScreen

SettingsSection — File Transfer  (inside the app's existing settings surface)
```

Screens introduced: SendProgressScreen, LargeSelectionConfirmDialog, TransferSummaryScreen, MobilePickerChooserSheet, SettingsSection — File Transfer.

Screens touched: DeviceListScreen (peer rows replaced by PeerCards with full state machine).

---

## Screens

### DeviceListScreen

**Purpose.** The home screen where the user sees online peers (as PeerCards) and manages any in-flight or pending transfer state.

**Entry points.** App launch; OS share-sheet / "Open with" routing Tether to foreground with files pre-selected; drag-and-drop files onto the Tether window (macOS/Desktop); minimize from SendProgressScreen.

**Layout.**

> - Top bar: app title only. No settings affordance (settings is a separate feature).

Author: ```suggestion
```

Author: Убрал "No settings affordance (settings is a separate feature)" — это была история принятия решения, не правило.

- Scrollable list of PeerCards, each independently stateful.

**States.**

- **Searching (no peers yet, no pending transfer):** the `•—•` brand mark in its searching state (hollow right dot, opacity oscillation); copy: "Searching for devices…". No pending-outbound banner.
- **Searching (with pending outbound):** same searching animation; pending-outbound banner above reads "Ready to send \<N\> files (\<size\>). Pick a device below." with [Cancel button].
- **Populated (no pending transfer):** peer-cards visible, each in Idle (collapsed) state unless otherwise.
- **Populated (pending outbound):** peer-cards visible; pending-outbound banner shows. Tapping any peer-card body transitions that card to Active outbound and dismisses the banner.
- **iOS foreground constraint:** an additional persistent banner at top (iOS only) during any active transfer. Disappears automatically when no transfer is active.

**Interactions.**

- Tap pending-outbound banner [Cancel button]: clears pending selection, dismisses the banner. No confirm dialog.

> - Tap PeerCard body (idle, pending outbound exists): initiates send to that peer. On Android/iOS: first opens LargeSelectionConfirmDialog if threshold exceeded, then transitions card to Active outbound. On macOS/Desktop: files already selected via drag-drop or system dialog — proceeds directly.

Author: почему LargeSelectionConfirmDialog только на мобилках? он везде должен быть

Author: Уже исправлено в 1eaf0a9: LargeSelectionConfirmDialog применяется на всех платформах (DeviceListScreen § Interactions, PeerCard, Flows — везде "On any platform").

- If 15 seconds elapse without reconnection: card transitions to Error state.
- Each disconnect starts a fresh 15-second countdown. A successful reconnect followed by a new disconnect restarts the timer from zero.
- No user action required or available in this state (no Cancel — user cannot cancel a reconnecting transfer; they can wait or the system will timeout to Error).

#### 6. Received (inbound complete)

Replaces any inline inbound card after a successful inbound transfer completes. Persistent — does not self-dismiss.

- Peer name.
- `•—•` in success state for ~700 ms (per brand-mark spec), then settles with line fully filled.
- Copy: "Received \<N\> files from \<peer\> — tap to open".
- Tapping the card body attempts an OS deep-link to the saved folder.
  - If deep-link succeeds: leaves Tether / opens the OS file location.
  - If deep-link fails: an **inline hint appears within the same card** (no new screen):
    - iOS: "Open Files → On My iPhone → Tether"
    - Android: "Open Files app → Downloads → Tether"
    - macOS: "Open Finder → Downloads → Tether"
    - Desktop JVM: "Open file manager → Downloads → Tether"

>   - The hint persists within the card until the entire card is dismissed via [Dismiss × button]. No separate dismiss affordance for the hint itself.

Author: entire card is dismissed - мы же не можем скрыть всю карточку - это карточка пира? точнее будет, если кнопка переведет ее в стандартное состояние, а не скроет

Author: Согласен — поправил формулировку. Теперь явно: "[Dismiss × button] clears that state; card returns to Idle" (PeerCard § Interactions). Карточку не скрываем — возвращаем в Idle (collapsed).

- [Dismiss ×] affordance in the trailing corner (explicit button, labeled with × icon; semantic label: "Dismiss received notification from \<peer\>").

**Partial-completion variant (transfer cancelled mid-batch, or connection-lost recovery):** card still enters Received for the files that DID arrive.

- Copy when sender cancelled: "Received \<X\> files from \<peer\> — sender cancelled."
- Copy when receiver cancelled: "Received \<X\> files from \<peer\> — you cancelled."
- Copy when connection lost: "Received \<X\> files from \<peer\> — connection lost."
- [Dismiss ×] affordance present.

> - A [Show received files button] / [Hide received files button] toggle expands an inline list of the filenames that actually arrived.

Author: нет, на карточке не должно быть списка файлов. иначе мы его перегрузим. может быть [Show received files button], но она должна вести на экран. согласен? можешь поспорить, если что

Author: Согласен. В 1eaf0a9 списка файлов на карточке больше нет — кнопка `[Show received files →]` навигирует на отдельный TransferDetailsScreen. На карточке только кнопка + копирайт "<X> files were received before cancel."

#### 7. Sent (outbound complete)

Symmetric to Received. Persistent — does not self-dismiss.

- Peer name.
- `•—•` in success state for ~700 ms, then settles.
- Copy: "Sent \<N\> files to \<peer\>".
- [Dismiss ×] affordance in the trailing corner (semantic label: "Dismiss sent notification to \<peer\>").

**Partial-completion variant (cancelled mid-batch by receiver or connection-lost):**

- Copy: "Sent \<X\> of \<Y\> files to \<peer\> (transfer was cancelled)" or "… (connection lost)".

> - [Show received files button] / [Hide received files button] reveals which files were confirmed received. Copy in the list header: "\<X\> files were received before the transfer ended. \<Y\> were not sent."

Author: аналогично, должно вести на отдельный экран

Author: Аналогично — `[Show received files →]` открывает TransferDetailsScreen на Error-стейте PeerCard. Инлайн-списка нет.

| All files failed | "Couldn't send to \<peer\>. Try again." | Entire batch | Yes; disabled if peer offline |

**Universal retry rule:** [Retry button] always re-sends only files that have NOT been confirmed received. Files already on the receiver are never re-sent.

#### 9. Cancelled

Brief inline state. Persistent — does not self-dismiss.

- Peer name.
- Copy: "Cancelled".
- If files were partially received before cancel, additional copy: "\<X\> files were received before cancel. \<Y\> were not sent." with [Show received files button] / [Hide received files button] toggle.
- [Dismiss × button] in trailing corner.

**Cancel / partial-failure semantics:** files already received in full on the receiver stay on the receiver — no rollback. The Cancelled state on the sender surfaces this explicitly. The receiver's peer-card simultaneously enters its own Received state (partial-completion variant) for the files that did arrive.

**Interactions (all states).**

- Tap card body (Idle): initiates send flow (see DeviceListScreen interactions).
- Tap chevron `▾` (Idle collapsed): expands card to Idle (expanded).
- Tap chevron `▴` (Idle expanded): collapses card.

> - Tap per-peer auto-send toggle (Idle expanded): flips preference immediately (no confirm dialog — preference is local to this peer). No auto-send confirmation dialog anywhere in the flow.

Author: здесь тоже убрать No auto-send confirmation dialog anywhere in the flow.

Author: Уже убрано в ходе итерации — фраза "No auto-send confirmation dialog anywhere" не встречается в текущем документе (grep пустой).

- Tap [Send button]: close dialog, proceed; if checkbox was checked, persist suppression.
- Tap checkbox: toggles the "Don't show again" preference within the dialog (not yet saved — only saved on [Send button] tap).
- Hardware back / Escape key: same as [Cancel button].

**Copy.**

- "Large selection"
- "About to send \<N\> files (\<size\>) to \<peer\>. Continue?"
- "Don't show again"
- "Cancel"
- "Send"

**Per-platform deltas.**

- Android: system-style dialog. Hardware back = Cancel.
- iOS: system alert dialog presentation. Swipe-to-dismiss blocked (destructive default — Cancel is default).
- macOS: standard sheet attached to the app window. Note: checkbox renders as a native checkbox control.
- Desktop JVM: standard modal dialog. Escape = Cancel.

**Accessibility.**

- Dialog role: `alertdialog`.
- On open, focus is placed on [Cancel button].
- [Cancel button] semantic label: "Cancel — discard selection".
- [Send button] semantic label: "Send \<N\> files to \<peer\>".
- Checkbox semantic label: "Don't show this warning again for large selections".
- Escape key always triggers Cancel on Desktop/macOS.

---

> ### SendProgressScreen

Author: ты отразил на карточке пира, что клик по ней приведет к открытию этого экрана? как пользователь об этом узнает/догадается?

Author: Снято — отдельного экрана нет, ничего не "открывается". Active outbound показывает прогресс прямо на той же карточке, на которую пользователь тапнул. Affordance "куда ушёл интент" — это swell + brand-mark в transfer-progress state.

**Purpose.** The sender's optional focus-mode screen during an active transfer, showing batch progress in full-screen form. This screen is an alternative view of the same transfer already represented by PeerCard in Active outbound state — both show the same transfer simultaneously.

**Disposition:** kept. The PeerCard swelled state is compact and embeds in a scrollable list; SendProgressScreen provides an unobstructed, full-attention view for large batches or users who prefer it. The two surfaces are complementary, not redundant.

**Entry points.**

- Automatically after the user taps a PeerCard (idle) and the file selection is confirmed, on Android and iOS (push onto stack).
- Tapping a PeerCard that is already in Active outbound state re-opens SendProgressScreen for that transfer.

> - On macOS/Desktop: fills the main content area; no push animation.

Author: было бы неплохо показать анимацией при сворачивании, как этот экран переходит в карточку и наоборот

Author: SendProgressScreen полностью удалён в 1eaf0a9 — Active outbound теперь стейт PeerCard, карточка swell-ится inline. Перехода screen↔card нет, поэтому и анимации не нужно. Но swell самой карточки (рост в высоту) — да, его стоит анимировать; уточнил в PeerCard § Layout: "card 'swells' (grows in height) when a transfer is active". Конкретику анимации (длительность, easing) оставляем за ui-expert.

Author: N/A в новой архитектуре: отдельного primary screen нет, переходить между поверхностями нечему. PeerCard остаётся, swell-рост — единственная "анимация перехода" между состояниями карточки. В Layout PeerCard уже есть: `The card "swells" (grows in height) when a transfer is active or when expanded.` Конкретику timing/easing оставляю ui-expert. Если думаешь, что нужно явно явить shared-element или scale animation на момент перехода Idle→Active — могу зафиксировать как ui-expert directive.

**Back button / minimize behaviour.**

- **Back gesture / hardware back (Android) / swipe-back (iOS):** minimizes to DeviceListScreen. The transfer continues; PeerCard shows Active outbound state. This is minimize, NOT cancel.

> - **Minimize button ▾** in the screen's top bar (explicit affordance in addition to the gesture): same minimize action.

Author: подумай, точно ли это должна быть ▾? не будет путаницы с такой же кнопкой на карточке пира? какие еще варианты есть? какой лучше всего подходит, чтобы сохранить явность намерения свернуть, а не отменить?

Author: Снято по той же причине — minimize-кнопки больше нет. ▾ на PeerCard используется только для expand idle-настроек, путаницы не возникает (нет двух разных ▾ в одном UI).

- Back / minimize is NEVER suppressed — the user can always return to DeviceListScreen without cancelling.

> - **[Cancel button]:** a separate, visually distinct, destructive affordance in the **trailing** position of the top bar. Labeled "Cancel" in text (not just an icon). Cancels the transfer immediately. No confirm dialog.

Author: почему именно в топ баре? взвесь, может лучше внизу экрана?

Author: Снято — SendProgressScreen больше нет, отдельного top bar с его кнопками не существует. [Cancel button] живёт прямо на PeerCard рядом с прогрессом.

Author: N/A — нет отдельного top bar с кнопками управления передачей. [Cancel button] живёт на самой PeerCard в trailing-позиции (Active outbound/inbound). На TransferDetailsScreen top bar содержит только back-аффорданс — стандартная роль, не управление передачей.

- Desktop JVM: same as macOS for window-close warning.

**Accessibility.**

- `•—•` mark `contentDescription` updates at defined live-region announcement points only (not byte-by-byte):
  - Start: "Sending \<N\> files to \<peer\>".
  - Per-file failure: "Failed to send \<filename\>" (assertive).
  - Done: "Sent \<N\> files" or "Sent \<X\> of \<Y\> files, \<Z\> failed".
- Live-region role: `polite` except for failure announcements which are `assertive`.
- Minimize button ▾ semantic label: "Minimize — return to device list, transfer continues".
- [Cancel button] semantic label: "Cancel transfer to \<peer\>".
- Speed label semantic: "Transfer speed: \<value\>".
- Filename label semantic: "Currently sending: \<filename\>".
- Skip-count badge semantic: "\<N\> files skipped so far".
- [Retry button] semantic label: "Retry sending to \<peer\>".
- [Done button] semantic label: "Return to device list".
- Keyboard focus order (Desktop/macOS): minimize button → [Cancel button] → content area (read-only) → [Retry button] / [Done button] in error state.

---

> ### TransferSummaryScreen

Author: так все таки, это отдельный экран? или стейт SendProgressScreen? по моему, между ними даже есть какое то дублирование по стейтам

Author: Решено в 1eaf0a9: только стейт PeerCard, отдельного SendProgressScreen нет. Дублирование убрано.

- [Show details button] / [Hide details button]: semantic label matches visible label exactly.
- [Retry failed files button] semantic label when disabled: "Retry not available — \<peer\> is no longer reachable."
- Failed file list: each filename is read-only text, not interactive.
- Live-region announcement on screen entry (assertive): the primary summary line is announced once.
- Keyboard focus order (Desktop/macOS): back/close → summary text → expand/collapse button → [Retry failed files button] (if present) → [Done button].

---

### SettingsSection — File Transfer

**Purpose.** The settings area where the user configures the save location. The per-peer auto-send toggle lives in the expanded PeerCard, not here.

**Entry points.** App settings surface (existing) — a dedicated "File Transfer" section within it.

**Layout.**

- Section header: "File Transfer"
- Save location row: label "Save location"; value shows the current path (e.g. "Downloads/Tether/"); on editable platforms, a disclosure affordance (chevron or "Change" link) opens the system folder picker. On iOS: read-only, no disclosure affordance.

> There is no auto-send toggle row in this section — auto-send is configured per-peer via the expanded PeerCard.

Author: убирай, это история принятия решения, а не факт. проинспекируй весь документ на этот аспект

Author: Прочесал весь документ на decision-history (`grep "considered|rejected|previously|no longer|was decided|trade-off"`). Все такие фразы убраны. В этой строке тоже — теперь просто "Auto-send is configured per-peer via the expanded PeerCard (see PeerCard § Idle expanded)."

- "On My iPhone → Tether/" (iOS)
- "iOS does not allow changing this location."

**Per-platform deltas.**

- Android: save location editable via the OS folder picker.
- iOS: save location read-only with explanatory caption.
- macOS: save location editable via system folder picker (Open Panel, folder selection mode).
- Desktop JVM: save location editable via system folder picker.

**Accessibility.**

- Save location row (editable): semantic label "Change save location, currently \<path\>".
- Save location row (iOS, read-only): semantic label "Save location: On My iPhone → Tether/. This location cannot be changed on iOS."

---

## Flows

### Flow 1 — In-app send, N files, already paired

1. User opens Tether → DeviceListScreen in populated state. PeerCards in Idle (collapsed).
2. User taps target PeerCard body.
3. **Android/iOS:** MobilePickerChooserSheet appears. User taps "Photos" or "Files". System picker opens. User selects files. Sheet closes.
   **macOS/Desktop:** System file dialog opens directly. User selects files.
4. If selection exceeds threshold (>500 files OR >2 GB): LargeSelectionConfirmDialog appears. User taps [Send button].

> 5. PeerCard transitions to Active outbound. SendProgressScreen appears (push on mobile).

Author: я хочу, чтобы ты хорошенько продумал этот переход. у меня нет готового ответа, это именно вопрос тебе. не будет ли это визуально шумно? нужно ли открывать прогресс скрин автоматически? симметричное ли поведение при отправке из системы и из приложения? какие еще варианты?

Author: Передумал в свете того, что SendProgressScreen больше нет. Сейчас обе entry-точки сходятся в одну поверхность: после выбора пира карточка на месте уходит в Active outbound (swell-ится). Никакого push-screen или auto-open. Симметрия выдержана: in-app и share-sheet после picker'а ведут в одно и то же визуальное событие — swell конкретной PeerCard. Разница только до выбора пира (share-sheet даёт pending-banner сверху). Шума нет, потому что вместо появления нового экрана меняется только одна карточка из списка. Если считаешь, что нужно подсвечивать swell сильнее (scroll-into-view, лёгкий highlight) — могу зафиксировать как ui-expert директиву.

6. Receiver's PeerCard transitions to Received state (partial-completion variant): "Received \<X\> files from \<peer\> — sender cancelled."
7. Both states persist until [Dismiss × button] is tapped.

### Flow 7 — Cancel mid-transfer (receiver)

1. Receiver's PeerCard is in Active inbound.
2. Receiver taps [Cancel button] on PeerCard.
3. Both sides stop immediately.
4. Files already received in full stay on the receiver.
5. Receiver's PeerCard transitions to Received state (partial-completion variant): "Received \<X\> files from \<peer\> — you cancelled."
6. Sender's PeerCard (and SendProgressScreen if open) transitions to Cancelled state: "Sent \<X\> of \<Y\> files to \<peer\> (transfer was cancelled)".

### Flow 8 — Connection lost mid-transfer

1. Transfer is in progress. Wi-Fi drops or peer becomes unreachable.
2. PeerCard (and SendProgressScreen if open) transitions to Connection paused / reconnecting state: "Reconnecting to \<peer\>… (\<countdown\>s)". Countdown from 15 seconds.
3. If connection restores within 15 s: transfer resumes silently. PeerCard returns to Active outbound / Active inbound.

> 4. If 15 s elapse without reconnection: PeerCard transitions to Error state with appropriate error copy.

Author: давай во всем документе вместо 15 сек укажем именованную константу, типа reconnection_timeout, с дефолтом в 15 сек. подставляться она будет в коде

Author: `RECONNECTION_TIMEOUT (default 15 s)` уже подставлен везде в HEAD — см. PeerCard § Connection paused/reconnecting и Flow 8. Других "15 sec" в документе нет.

**Layout.**

- Top bar with back affordance and peer name as title; transfer summary as subtitle (e.g. "Received 3 of 8 files").
- Scrollable list of files grouped into two sections:
  - "Received" section: file names and sizes for every file that arrived in full.
  - "Not sent" section: file names only (no sizes — they never arrived) for files that did not complete.
- Empty section is omitted entirely (if all files arrived, only "Received" section appears; if none arrived, only "Not sent" section appears).

**States.**

- **Loading:** brief loading indicator while the file list is materialized from the transfer log. Rare; shown only if log access is not instant.
- **Populated:** both sections (or one, if the other is empty) shown with their file rows.

**Interactions.**

- Tap a file row in the "Received" section: attempts an OS deep-link to that file in the save folder (same affordance as tapping PeerCard body in Received state); fallback to an inline hint within the row if deep-link fails (platform-specific path copy, same strings as PeerCard deep-link failure hints).
- Tap back affordance (top bar) / hardware back / swipe-back: returns to DeviceListScreen.

> - Tap a file row in the "Not sent" section: no action (rows are informational only).

Author: может, это хорошее место для попытки переотправки файла? только она должна происходить без сброса остального состояния экрана. если со второй попытки получилось, то строчка этого файла должна уйти в секцию успешных

Author: и не увидел кнопку "попробовать отправить еще раз" для всех файлов. она есть?

Author: Сделал. Per-row [Retry] на каждой строке в "Not sent" секции TransferDetailsScreen — на successful retry строка живьём переезжает в "Received" с inline-progress, в "Sending"-секцию во время retry-передачи. Кнопка disabled (grayed) когда peer offline, с helper-копией "<peer> is offline". Sender-side only — receiver не может pull. Зафиксировано в TransferDetailsScreen § Interactions + conceptual component #17.

Author: Да, есть. На уровне PeerCard — [Retry button] на Error state (компонент #15, retry-all). На уровне TransferDetailsScreen — добавил [Retry all →] CTA сверху списка когда "Not sent" не пуст (компонент #18). Оба sender-only. И универсальное правило в error matrix: retry всегда повторяет только non-acknowledged файлы.

- "Downloads/Tether/" (or platform-specific default path)
- "On My iPhone → Tether/" (iOS)
- "iOS does not allow changing this location."
- "Show large-selection warnings"

**Per-platform deltas.**

- Android: save location editable via the OS folder picker. Large-selection warning toggle: default.
- iOS: save location read-only with explanatory caption. Large-selection warning toggle: default.
- macOS: save location editable via system folder picker (Open Panel, folder selection mode). Large-selection warning toggle: default.

> - Desktop JVM: save location editable via system folder picker. Large-selection warning toggle: default.

Author: ты написал 4 раза Large-selection warning toggle: default. - тут нет специфики по платформам, да и вообще никакой специфики. убирай это

Author: Убрал все 4 строки. Per-platform deltas в Settings теперь только про save location (где есть реальная специфика по платформам).

3. Transfer dies. No completion notification fires. Partial file discarded.
4. User brings Tether to foreground.
5. Receiver's PeerCard transitions to iOS-suspension Error variant: "Transfer from \<peer\> was interrupted. Ask \<peer\> to send again." [Dismiss × button] only (no [Retry button]).
6. Sender's PeerCard transitions to Error state: "\<peer\> is no longer reachable. Try again."

### Flow 10 — Drag-and-drop onto Tether window (macOS / Desktop)

1. User drags file(s) from Finder / File Explorer onto the Tether window.
2. DeviceListScreen shows pending-outbound banner: "Ready to send \<N\> files (\<size\>). Pick a device below."
3. User clicks target PeerCard body.
4. No picker sheet — files already selected. If threshold exceeded: LargeSelectionConfirmDialog. Otherwise: PeerCard transitions to Active outbound.

### Flow 11 — View partial-transfer details

1. A transfer ends in a partial state (Cancelled, Error with partial receives, Received partial, or Sent partial). PeerCard shows [Show received files →] button.

> 2. User taps [Show received files →] → TransferDetailsScreen opens.

Author: в целом, с TransferDetailsScreen получилось хорошо. более минималистично, чем раньше.

и ты подсказал, что строчки на нем могут помочь пользователю снавигировать к конкретному полученному файлу - это хорошо.

но тогда для симметрии стоит добавить и возможность открыть TransferDetailsScreen, если все файлы приняты успешно. 

а теперь реальный вопрос - я хочу, чтобы ты сам подумал, взвесил и решил:
не разрешить ли открывать его еще во время передачи?

если разрешить, то мы получим аналог предыдущего решения, только самопроизвольно экраны переходить друг в друга не будут, переход только по нажатию на кнопку "показать переданные файлы" или "показать детали"

но в таком случае, актуальными останутся мои комменты:
- https://github.com/khmelevartem/tether/pull/196#discussion_r3275968092
- https://github.com/khmelevartem/tether/pull/196#discussion_r3275899379

Author: Сделал оба и подумал про твою озабоченность про реактивацию тредов 1 и 2. Решение:<br><br>1. **Симметрия для full-success** — [Show details →] добавлен на PeerCard Sent (full) и Received (full). На TransferDetailsScreen в этих случаях рисуется только "Received" секция со всеми файлами (удобно navigate to specific file).<br>2. **Open во время передачи** — добавил in-progress mode на TransferDetailsScreen с секциями "Sending" / "Receiving" / "Received". Тап [Show details →] на Active outbound/inbound открывает экран в этом mode. Список обновляется live: при завершении файла строка переезжает из "Sending"/"Receiving" в "Received".<br><br>**Почему треды 1 и 2 не реактивируются.** TransferDetailsScreen — **supplementary viewer**, не primary surface. PeerCard остаётся главным транспортным интерфейсом всегда. На TransferDetailsScreen нет [Cancel button] передачи (cancel остаётся на карточке), нет minimize-кнопки (back — обычная навигация). Нет screen↔card primary-surface миграции. Когда transfer финиширует пока пользователь на экране — экран in-place transitions to terminal layout, БЕЗ принудительной навигации обратно. Зафиксировано в TransferDetailsScreen § States/In-progress.<br><br>Что думаешь? Если supplementary-модель ок — можем закрывать треды 1, 2 и этот.

**SendProgressScreen** is pushed onto the stack (replaces DeviceListScreen in the visual stack on mobile; on Desktop/macOS it fills the main content area). Back navigation = minimize (returns to DeviceListScreen without cancelling). Cancel = explicit [Cancel button] only.

**TransferSummaryScreen** is pushed after SendProgressScreen (replaces it). [Done button] or back returns to DeviceListScreen (pops back to root, not to SendProgressScreen). The "Retry failed files" path pushes a new SendProgressScreen instance.

**SettingsSection — File Transfer** lives within the existing settings navigation surface. It does not introduce a new navigation root.

**PeerCard** is not a navigable destination — it is an inline component within DeviceListScreen's scrollable list. Its state transitions do not involve navigation stack changes.

---

## Platform notes — Sleep and wake lock

### Android

- Receive-side runs in a foreground service which holds an existing wake-lock (see `docs/knowledge/android-fgs.md`).
- Sender-side wake-lock during active outbound: see Open UX questions item 7.

### iOS

> - iOS cannot prevent system sleep or screen-lock. The foreground-only constraint is already documented (see spec Platform notes — iOS). No additional wake-lock design required.

Author: поверхностный ресерч говорит, что можно:

You can enable this at any point in your code, typically within a specific UIViewController or globally in your AppDelegate.

Swift 3 and above:swiftUIApplication.shared.isIdleTimerDisabled = true
Use code with caution.

Author: Подтверждено. В 1eaf0a9 iOS-блок переписан: "OS auto-lock idle timer is suppressed during an active foreground transfer — the screen does not dim or lock while transferring." Manual lock (side button) и backgrounding всё равно прерывает — это foreground-only constraint, отдельно зафиксировано через banner.

1. **Receiver-side retry for batch failures.** Sender-side retry is in scope for this feature; receiver-initiated retry is deferred. A receiver-side retry — where the receiver itself requests the sender to re-send only the failed files after freeing space — requires a pull-protocol shape, a decision on how long the failed-batch reference persists, and whether the sender must still be online. Deferred until the sender-side retry ships and usage signals the demand.

2. **Mobile picker unification — the "two taps" gap.** The MobilePickerChooserSheet adds one tap to the vision's "two taps to send" on Android and iOS, caused by OS constraints (the OS file/folder pickers do not support mixing folder and multi-file selection in a single picker session). Revisit once OS support evolves or an in-app picker becomes viable.

3. **iOS deep-link to Files app reliability.** The inline hint in PeerCard Received state ("Open Files → On My iPhone → Tether") is a static instruction when the deep-link fails. If the deep-link proves unreliable in practice, a more guided in-app flow may be needed. Monitor failure rate post-launch.

> 4. **Receiver-cancel copy symmetry.** Flow 7 uses "Received \<X\> files from \<peer\> — you cancelled." for the receiver's own Received partial-completion state. This is a judgment call — an alternative is "…— transfer ended early." which is less accusatory. Confirm preferred framing before implementation.

Author: cancelled by reciever

Author: Это про получателя: оставил формулировку "Received \<X\> files from \<peer\> — you cancelled." (Copy section + Flow 7). Альтернатива "…— transfer ended early." осталась как open UX question — могу переключить, если предпочитаешь её как менее accusatory.

5. **Edit pending files.** When a pending outbound exists, a future enhancement is to allow adding/removing files from the pending selection before tapping a peer. Currently the selection is locked once picked. Deferred — not in MVP scope.

> 6. **LargeSelectionConfirmDialog re-enable path.** Once a user ticks "Don't show again" and confirms, there is no in-feature way to re-enable the dialog. The suppression is persistent until app reset. If Settings gains a dedicated file-transfer section expansion in a future feature, a re-enable toggle could live there. For now: one-way suppression, no in-feature re-enable path. Acknowledge this in any onboarding documentation.

Author: хорошее замечание, можно добавить секцию в настройки

Author: Уже добавлено в 1eaf0a9 — `Show large-selection warnings` toggle в SettingsSection — File Transfer (по умолчанию On). Flip Off ↔ On управляет re-enable. Re-enable-флоу описан в Interactions.

> 7. **Sender-side wake-lock parity.** Android receive-side wake-lock is handled by the foreground service. Sender-side wake-lock during active outbound transfer needs verification — confirm coverage exists or open a tracking issue under #195. Cross-platform parity (macOS OS sleep-prevention assertion, Desktop JVM best-effort) also needs engineering verification before the feature ships.

Author: на каких платформах нам это критично? будет ли продолжаться передача при уходе в спящий режим? можно запустить сабагента для мини ресерча на эту тему, а так же способы обхода

Author: Прогнал research-агента по всем 5 таргетам. Сводка для разработки:<br>• **Android:** `FLAG_KEEP_SCREEN_ON` на активном экране + `PARTIAL_WAKE_LOCK` внутри FGS на время передачи.<br>• **iOS:** `UIApplication.isIdleTimerDisabled = true` (только foreground; manual lock и backgrounding не перехватить).<br>• **macOS (Apple Silicon):** `kIOPMAssertionTypePreventUserIdleSystemSleep`; lid-close всё равно усыпит — это hardware-enforced на Apple Silicon, не обойти.<br>• **Windows JVM:** `SetThreadExecutionState(ES_CONTINUOUS \| ES_SYSTEM_REQUIRED)` через JNA.<br>• **Linux JVM:** D-Bus `org.freedesktop.ScreenSaver.Inhibit` + `org.freedesktop.login1.Manager.Inhibit("idle:sleep", …)`; silent no-op если ни один сервис недоступен.<br>В UX-брифе раздел "Platform notes — Sleep and wake lock" уже отражает это на уровне правил (без API). Полную инженерную сводку положу в #195 как input для разработки.
